fast64_internal/sm64/sm64_objects.py

```diff
@@ -719,11 +726,80 @@ def to_c(self):
             + str(int(round(self.camFocus[1])))
             + ", "
             + str(int(round(self.camFocus[2])))
-            + "},"
+            + "}"
         )
         return data
 
 
+@dataclasses.dataclass
+class CustomCmd:
+    cmd: str | int
+    transform: mathutils.Matrix
+    whole_matrix: bool
+    include_trans: bool = False
+    rot_type: str = "None"
+    include_scale: bool = False
+    parameter: str | None = None
+    color: list[int] | None = None
+
+    name: str = ""  # for sorting
+    hasDL: bool = False
+
+    @property
+    def matrix(self):
+        return [round(column, 4) for row in self.transform for column in row] if self.whole_matrix else None
+
+    @property
+    def trans(self):
+        return [round(x, 4) for x in self.transform.to_translation()] if self.include_trans else None
+
+    @property
+    def rot(self):
+        if self.rot_type == "None":
+            return None
+        elif self.rot_type == "Euler":
+            return [round(math.degrees(x), 4) for x in self.transform.to_euler()]
+        elif self.rot_type == "Quaternion":
+            return [round(x, 4) for x in self.transform.to_quaternion()]
+        elif self.rot_type == "Axis Angle":
+            axis_angle = self.transform.to_quaternion().to_axis_angle()
+            return [round(x, 4) for x in axis_angle[0]] + [round(axis_angle[1], 4)]
+        assert False, f"Unknown rotation type {self.rot_type}"
+
+    @property
+    def scale(self):
+        return [round(x, 4) for x in self.transform.to_scale()] if self.include_scale else None
+
+    def size(self):
+        return 8
+
+    def to_binary(self, segmentData):
+        raise PluginError("Custom commands are not supported for binary exports.")
+
+    def to_c(self, depth=0):
+        assert isinstance(self.cmd, str), "Command is not str"
+        arg_groups = []
+        if self.matrix is not None:
+            arg_groups.append(f"/*matrix*/ {', '.join([f'{x}f' for x in self.matrix])}")
+        else:
+            if self.trans is not None:
+                arg_groups.append(f"/*trans*/ {', '.join([f'{x}f' for x in self.trans])}")
+            if self.rot is not None:
+                arg_groups.append(f"/*rot ({self.rot_type.lower()})*/ {', '.join([f'{x}f' for x in self.rot])}")
+            if self.scale is not None:
+                arg_groups.append(f"/*scale*/ {', '.join([f'{x}f' for x in self.scale])}")
+        if self.color is not None:
+            arg_groups.append(f"/*color*/ {', '.join([str(x) for x in self.color])}")
+        if self.parameter is not None:
+            arg_groups.append(f"/*param*/ {self.parameter}")
+        if len(str(arg_groups)) > 100:
+            seperator = ",\n" + ("\t" * (depth + 1))
```

Comment:
it makes no sense to add a `_depth` arg to all the associated macro functions and then only use it in this one specific instance. There likely is a better way of determining the proper amount of tabs to be used.

Reply:
Open to suggestions, I'm not checking  the function signature that's for sure

Reply:
is there ever a case in which the depth is different than 1 and could not be adjusted for in the calling function?
you can return unformatted f strings, e.g. a "{tabs} macro(args)" then run `str.format` on the returned value so that it inserts the tabs later, allowing you to defer indentation until the root caller.
Imo this is better than carrying args to full depth

Reply:
I think formatting every string for depth is much more hacky than passing in depth, are you sure about this?